src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/TrieBuilder.cs

```diff
@@ -0,0 +1,555 @@
+            TrieBuilder builder = new TrieBuilder(regexNode);
+            NodeCollection matchNodes = builder.Add(new NodeCollection(TrieNode.Root), regexNode, isFinal: true, out _);
+            builder.AcceptMatches(matchNodes);
+
+            // If we still have one trie node -the root one-, the regex node has no fixed leading part
+            // (for example it starts with (a)*) and we cannot use the trie.
+            // Furthermore if the root node is a match, it means that a regex may start with something not
+            // in the trie (for example in a*|b|c, the trie will have b and c but not a) so we cannot use it either.
+            if (builder._nodes.Count == 1 || builder._nodes[TrieNode.Root].IsMatch)
+            {
+                return null;
+            }
+            builder.ValidateInvariants();
+            return builder._nodes;
+        }
+
+        /// <summary>
+        /// Marks all nodes in <paramref name="nodes"/> as matching.
+        /// </summary>
+        private void AcceptMatches(NodeCollection nodes)
+        {
+            int count = nodes.Count;
+            for (int i = 0; i < count; i++)
+            {
+                _nodes[nodes[i]].IsMatch = true;
+            }
+        }
+
+        /// <summary>
+        /// A helper method that applies a one-to-one transformation in a <see cref="NodeCollection"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of <paramref name="state"/>.</typeparam>
+        /// <param name="nodes">The input <see cref="NodeCollection"/>.</param>
+        /// <param name="state">A parameter passed to <paramref name="fAdd"/>.</param>
+        /// <param name="fAdd">A function that accepts this trie, a node index of <paramref name="nodes"/>,
+        /// <paramref name="state"/>, and returns a new node index. Typically it would call either
+        /// <see cref="Add(ref TrieBuilder, int, char, bool, out bool)"/> or <see cref="Add(ref TrieBuilder, int, string, bool, out bool)"/>.</param>
+        /// <param name="isFinal">Whether we are at the end of the traversal algorithm.</param>
+        /// <param name="canContinue">Returns whether the traversal algorithm can continue past this transformation.</param>
+        private NodeCollection AddOneToOneHelper<T>(NodeCollection nodes, T state, OneToOneNodeMapping<T> fAdd, bool isFinal, out bool canContinue)
+        {
+            int count = nodes.Count;
+            switch (count)
+            {
+                case 0:
+                    canContinue = false;
+                    return NodeCollection.Empty;
+                case 1:
+                    int newNode = fAdd(ref this, nodes[0], state, isFinal, out canContinue);
+                    return new NodeCollection(newNode);
+                default:
+                    canContinue = false;
+#if REGEXGENERATOR
+                    HashSet<int> visited = new HashSet<int>();
+#else
+                    HashSet<int> visited = new HashSet<int>(count);
+#endif
```

Comment:
```suggestion
                    HashSet<int> visited = new HashSet<int>(
#if !REGEXGENERATOR
                       count
#endif
                    );
```

Reply:
It fails with `error SA1112: (NETCORE_ENGINEERING_TELEMETRY=Build) Closing parenthesis should be on line of opening parenthesis`. I will just create the hash set without the initial size.